Feature/co 440 class search

floppyears opened the pull request.

floppyears committed:
Renaming skeleton references

CO-440

floppyears committed:
Add dependencies

CO-440

floppyears committed:
Set configuration file

CO-440

floppyears committed:
Connect to backend host and spit data out

CO-440

This revision just connects to the backend host, authenticates and
returns the data out. It validates the parameters. Still left to do:
tests, parsing data out and adding jsonapi links.

floppyears committed:
Performs class search and uses jsonapi

CO-440

floppyears committed:
Update swagger

CO-440

floppyears committed:
Expanding text in a swagger field

CO-440

floppyears committed:
Flushing out the DAO

CO-440

Adding faculty and meeting times
Being specific about the type of Attribute fields to force errors

floppyears committed:
Update configuration example

CO-440

floppyears committed:
Misc updates and fixes

CO-440

Update readme
Cleaning up DAO

floppyears committed:
Add common jsonapi classes to skeleton

CO-440

Our api standards follow jsonapi.org. These classes represent the
properties common to most of our apis.

jaredkosanovic commented on src/main/groovy/edu/oregonstate/mist/classsearchapi/dao/ClassSearchDAO.groovy:
```diff
@@ -0,0 +1,184 @@
+                    creditHourLow:              it.creditHourLow,
+                    creditHours:                it.creditHours,
+                    enrollment :                it.enrollment,
+                    maximumEnrollment :         it.maximumEnrollment,
```
Is there a reason for the space before the colon? Several other attributes that are like this

floppyears replied:
No, my bad. I will update it.

floppyears committed:
Merge pull request #29 from osu-mist/feature/CO-440-jsonapi

Add common jsonapi classes to skeleton

floppyears committed:
Merge remote-tracking branch 'skeleton/master' into feature/CO-440-class-search

floppyears committed:
Minor fix from code review

CO-440

floppyears committed:
Add jsonapi pagination

CO-440

floppyears committed:
Add repeated logic to skeleton

CO-440

This will help with pagination of jsonapi.org apis. The actual
implementation will be different depending on the backend, but that
is up to the child Resource class.

floppyears committed:
Merge branch 'skeleton-master' into feature/CO-440-class-search

floppyears committed:
Move from http builder to apache httpclient

CO-440

Use httpclient from apache since it's built-in by dropwizard

floppyears committed:
Refactor dao for reusability

CO-440

floppyears committed:
Refactoring class search

CO-440

Making code modular for re-usability.
Added healthcheck

magical commented on readme.md:
```diff
@@ -1,6 +1,6 @@
+# Class Search API
 
-Skeleton for Dropwizard Web APIs.
+Class Search API to query the course catalog real time for course information and availability.
```
s/real time/in real time/

magical commented on src/main/groovy/edu/oregonstate/mist/api/Resource.groovy:
```diff
@@ -83,4 +105,90 @@ abstract class Resource {
+                                              MultivaluedMap<String, String> queryParameters) {
+        for (Map.Entry<String, List<String>> entry : queryParameters.entrySet()) {
+            // not an array parameter
+            if (!entry.key.contains("[") && !entry.key.contains("]")) {
```
should be ||

magical commented on src/main/groovy/edu/oregonstate/mist/api/Resource.groovy:
```diff
@@ -83,4 +105,90 @@ abstract class Resource {
+                    return entry.value?.get(0)
+                }
+            }
+        }
```
Why do you have to iterate over the entire map? Can't you just do queryParameters.getFirst(key+'['+value+']')? For that matter, can't callers just do uriInfo.getQueryParameters().getFirst('page[number'])?

magical replied:
getArrayParameters will also erroneously match a key with garbage after it. e.g., 'page[number]asdfjkh'

tsoliangwu0130 replied:
hmm, I think using "a" and "b" as variable names are bit ambiguous. How about using "leftBrketIdx" and "rghtBrketIdx" in here?

floppyears replied:
I agree with the comments. At first I was trying to write something generic, but after thinking about it more, I don't think that's needed. This code comes from the skeleton, so I updated that pull request: https://github.com/osu-mist/web-api-skeleton/pull/32/files

tsoliangwu0130 commented on src/main/groovy/edu/oregonstate/mist/classsearchapi/dao/ClassSearchDAO.groovy:
```diff
@@ -0,0 +1,203 @@
+                    faculty:                    faculty,
+                    meetingTimes:               meetingTimes
+            )
+
```
There are also some spaces before ":"

floppyears committed:
Misc code review fixes

CO-440

magical commented on src/main/groovy/edu/oregonstate/mist/api/Resource.groovy:
```diff
@@ -83,4 +105,90 @@ abstract class Resource {
+        nonNullParams.remove('pageNumber')
+
+        // remove empty GET parameters
+        nonNullParams = nonNullParams.findAll { it.value } .collect { k, v -> "$k=$v" }
```
the names and values should be urlencoded. does java not have a class for constructing urls?

magical commented on src/main/groovy/edu/oregonstate/mist/classsearchapi/resources/ClassSearchResource.groovy:
```diff
@@ -0,0 +1,104 @@
+import javax.ws.rs.core.MediaType
+
+/**
+ * Sample resource class.
```
stale comment

magical commented on src/main/groovy/edu/oregonstate/mist/classsearchapi/resources/ClassSearchResource.groovy:
```diff
@@ -0,0 +1,104 @@
+            setPaginationLinks(response.sourcePagination, term, subject, courseNumber, q, resultObject)
+
+            ResponseBuilder responseBuilder = ok(resultObject)
+            responseBuilder.build()
```
ok(resultObject).build()

magical commented on src/main/groovy/edu/oregonstate/mist/classsearchapi/dao/UtilHttp.groovy:
```diff
@@ -0,0 +1,92 @@
+        }
+
+        URI uri = uriBuilder.build()
+        uri
```
uriBuilder.build()

magical commented on src/main/groovy/edu/oregonstate/mist/classsearchapi/resources/ClassSearchResource.groovy:
```diff
@@ -0,0 +1,104 @@
+    }
+
+    private void setPaginationLinks(def sourcePagination, String term, String subject, String courseNumber, String q,
+                                    ResultObject resultObject) {
```
It's a little odd to have resultObject be an "out" parameter. You only set properties on resultObject.links. Can you just return a map?

magical commented on src/main/groovy/edu/oregonstate/mist/classsearchapi/dao/ClassSearchDAO.groovy:
```diff
@@ -0,0 +1,203 @@
+            totalCount: headerMap['X-hedtech-totalCount']?.toInteger(),
+            pageOffset: headerMap['X-hedtech-pageOffset']?.toInteger(),
+            pageMaxSize: headerMap['X-hedtech-pageMaxSize']?.toInteger()
+        ]
```
is headerMap really necessary?

floppyears replied:
I didn't have it there before, but it was such a pain to access the keys/values due to how the headers were returned. If it wasn't because of the features that the apache httpbuilder library provides and the metrics that dropwizards bake in I would switch back to the groovy class. It was far fewer lines of code and time debugging to use groovy.

magical replied:
Fair enough.

floppyears committed:
Misc code review fixes

CO-440
